### PR TITLE
Fix hostname related issues (HZ-222)

### PR DESCRIPTION
**Pull request by jbartok**
Fix for issues described in [Analyse and fix issues with using hostnames in member and WAN configuration](https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/2347827205/Analyse+and+fix+issues+with+using+hostnames+in+member+and+WAN+configuration).

Fixes [HZ-222](https://hazelcast.atlassian.net/browse/HZ-222)

The test scenarios that have been used to confirm both the problems and the fix are described in [this document](https://docs.google.com/document/d/1q2w-Ya4vRsWjy5M9Ww5HbYJ1mT3TVRcIevRylZsNiwU/edit?usp=sharing). Atm. I'm unable to automate these tests.

### CODE REVIEW

**Review by kwart — COMMENTED** [review]
One of the simple scenarios is the following:

```java
    @Test
    public void test() {
        Config config = new Config();
        JoinConfig join = config.getNetworkConfig().getJoin();
        join.getMulticastConfig().setEnabled(false);
        join.getTcpIpConfig().setEnabled(true).clear().addMember("127.0.0.1");
        HazelcastInstance hz1 = Hazelcast.newHazelcastInstance(config);

        Config config2 = new Config();
        JoinConfig join2 = config2.getNetworkConfig().getJoin();
        join2.getMulticastConfig().setEnabled(false);
        join2.getTcpIpConfig().setEnabled(true).clear().addMember("localhost");
        HazelcastInstance hz2 = Hazelcast.newHazelcastInstance(config2);

        assertClusterSize(2, hz1, hz2);
    }
```

The current changes don't seem to fix this.

**Comment by jbartok** [comment]
> One of the simple scenarios is the following:
> 
> ```java
>     @Test
>     public void test() {
>         Config config = new Config();
>         JoinConfig join = config.getNetworkConfig().getJoin();
>         join.getMulticastConfig().setEnabled(false);
>         join.getTcpIpConfig().setEnabled(true).clear().addMember("127.0.0.1");
>         HazelcastInstance hz1 = Hazelcast.newHazelcastInstance(config);
> 
>         Config config2 = new Config();
>         JoinConfig join2 = config2.getNetworkConfig().getJoin();
>         join2.getMulticastConfig().setEnabled(false);
>         join2.getTcpIpConfig().setEnabled(true).clear().addMember("localhost");
>         HazelcastInstance hz2 = Hazelcast.newHazelcastInstance(config2);
> 
>         assertClusterSize(2, hz1, hz2);
>     }
> ```
> 
> The current changes don't seem to fix this.

Yes, this is the other, INDEPENDENT issue I've mentioned and that I'm tackling next. The test cases for it are [here](https://docs.google.com/document/d/1oIuXGtAEUhsD-W6X73CGmWaaarpLONBawSb5NuRoFyE/edit?usp=sharing).

**Comment by jbartok** [comment]
run-lab-run

**Comment by jbartok** [comment]
Shutdown issues apparently solved, but still want to further review it. Commiting to see if other tests fail.

File: hazelcast/src/test/java/com/hazelcast/cluster/TcpIpHostnameJoinTest.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    @BeforeClass
+    public static void beforeClass() throws IOException {
+        mockStatic(NetworkInterface.class);
+
+        List<NetworkInterface> networkInterfaces = new ArrayList<NetworkInterface>();
+        networkInterfaces.add(createNetworkConfig("lo", true, LOCALHOST, LOCALHOST2, LOCALHOST3, LOCALHOST4));
+        when(NetworkInterface.getNetworkInterfaces()).thenReturn(enumeration(networkInterfaces));
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/cluster/TcpIpHostnameJoinTest.java:77]:
```suggestion
        when(NetworkInterface.getNetworkInterfaces()).then(invocation -> enumeration(networkInterfaces));
```

to return a new enumeration each time `getNetworkInterfaces()` is called.

**Comment by jbartok** [comment]
run-lab-run

**Comment by jbartok** [comment]
run-lab-run

**Comment by jbartok** [comment]
run-lab-run

**Comment by kwart** [comment]
One of new tests still fails:
```
mvn test -Dtest=TcpIpHostnameJoinTest -Pnightly-build
...
java.lang.AssertionError: Cluster size is not correct. Expected: 2, actual: 1, instance index: 0
	at org.junit.Assert.fail(Assert.java:89)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:1047)
	at com.hazelcast.cluster.TcpIpHostnameJoinTest.test_whenMembersSpecifiedViaHostnames(TcpIpHostnameJoinTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.internal.runners.TestMethod.invoke(TestMethod.java:68)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runTestMethod(PowerMockJUnit44RunnerDelegateImpl.java:326)
	at org.junit.internal.runners.MethodRoadie$2.run(MethodRoadie.java:89)
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:97)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.executeTest(PowerMockJUnit44RunnerDelegateImpl.java:310)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTestInSuper(PowerMockJUnit47RunnerDelegateImpl.java:131)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.access$100(PowerMockJUnit47RunnerDelegateImpl.java:59)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner$TestExecutorStatement.evaluate(PowerMockJUnit47RunnerDelegateImpl.java:147)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.evaluateStatement(PowerMockJUnit47RunnerDelegateImpl.java:107)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTest(PowerMockJUnit47RunnerDelegateImpl.java:82)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runBeforesThenTestThenAfters(PowerMockJUnit44RunnerDelegateImpl.java:298)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:50)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.invokeTestMethod(PowerMockJUnit44RunnerDelegateImpl.java:218)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.runMethods(PowerMockJUnit44RunnerDelegateImpl.java:160)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$1.run(PowerMockJUnit44RunnerDelegateImpl.java:134)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:34)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:44)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.run(PowerMockJUnit44RunnerDelegateImpl.java:136)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.run(JUnit4TestSuiteChunkerImpl.java:117)
	at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMockRunner.run(AbstractCommonPowerMockRunner.java:57)
	at org.powermock.modules.junit4.PowerMockRunner.run(PowerMockRunner.java:59)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

**Comment by jbartok** [comment]
run-lab-run

**Review by vbekiaris — APPROVED** [review]
The changes look OK as an incremental fix.

That said, I am not happy with having code to deal with member identity issues spread across the codebase. At some point we should clean this area.

**Comment by jbartok** [comment]
run-lab-run

**Comment by jbartok** [comment]
run-lab-run

**Review by kwart — APPROVED** [review]

**Comment by kwart** [comment]
run-lab-run

**Comment by jbartok** [comment]
image

**Pull request merged**